### PR TITLE
Extended API Support

### LINKED ISSUE #16: Validate tools file
against a json schema and produce meaningful error messages.

For example, when skipping the `context` field, the server will silently fail.

### LINKED ISSUE #19: Inline Context for GenericChatMessage
when storing and retrieving the messages to make it easier to support different backend implementations like jdbc.

### PR DESCRIPTION
**Pull request by mbroecheler**
- support REST APIs with path and query parameters
- support multiple APIs with auth configuration
- support local functions with 
- added examples and made them all use publicly accessible APIs (moving DataSQRL specific examples to separate repo)
- fixes #19 
- fixes #16

### CODE REVIEW

File: java/apirag-core/src/test/resources/nutshop-c360.tools.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [
   {
-    "type": "graphql",
+    "type": "api",
```
Comment (bilbo7833) [inline java/apirag-core/src/test/resources/nutshop-c360.tools.json:3]: Why are graphql APIs just called `api`?

File: java/apirag-udf/src/test/java/com/datasqrl/ai/function/TestFunctionConversion.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import com.datasqrl.ai.function.builtin.CurrentTime;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.time.Instant;
+import java.util.Set;
+import lombok.SneakyThrows;
+import org.junit.jupiter.api.Test;
+import org.reflections.Reflections;
+
+public class TestFunctionConversion {
```
Comment (bilbo7833) [inline java/apirag-udf/src/test/java/com/datasqrl/ai/function/TestFunctionConversion.java:20]: Shouldn't this be called `FunctionConversionTest`?

**Review by bilbo7833 — CHANGES_REQUESTED** [review]
Looks good, but it is based on main before the Google support, which extends the Message Schema saved to the Backend. Shouldn't the Google Vertex support PR have been merged first?

File: java/apirag-core/src/test/resources/nutshop-c360.tools.json (hunk 1 above)
Comment (mbroecheler) [reply java/apirag-core/src/test/resources/nutshop-c360.tools.json:3]: Because we now support graphql and rest apis via different executors. Hence, I moved what kind of API it is into the API configuration (in the main configuration file).
We now just need to distinguish between functions that are called through an API vs functions that are called locally vs functions that are passed to the frontend. Those are the 3 "types".

**Comment by mbroecheler** [comment]
> Looks good, but it is based on main before the Google support, which extends the Message Schema saved to the Backend. Shouldn't the Google Vertex support PR have been merged first?

Yes, we'll rebase before the merge. Both of these PRs require that we update the examples. In this PR, all examples are using publicly available APIs. The DataSQRL examples are moved to datasqrl-examples and there is a separate PR for updating those. All of these need to be merged together.

File: java/apirag-udf/src/test/java/com/datasqrl/ai/function/TestFunctionConversion.java (hunk 2 above)
Comment (mbroecheler) [reply java/apirag-udf/src/test/java/com/datasqrl/ai/function/TestFunctionConversion.java]: Correct and fixed.

File: java/apirag-core/src/main/java/com/datasqrl/ai/backend/FunctionType.java [hunk 3]
```diff
@@ -5,7 +5,7 @@
 @Getter
 public enum FunctionType {
 
-  graphql(false),
+  api(false),
```
Comment (bilbo7833) [inline java/apirag-core/src/main/java/com/datasqrl/ai/backend/FunctionType.java:8]: @mbroecheler What is the difference here between type `api` and type `rest`?

File: java/apirag-core/src/main/java/com/datasqrl/ai/models/AbstractModelConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+public abstract class AbstractModelConfiguration implements ModelConfiguration {
+
+  public static final String MODEL_NAME_KEY = "name";
+  public static final String MAX_INPUT_TOKENS_KEY = "max_input_tokens";
+  public static final String MAX_OUTPUT_TOKENS_KEY = "max_output_tokens";
+  public static final String TEMPERATURE_KEY = "temperature";
+  public static final double TEMPERATURE_DEFAULT = 0.7;
+  public static final String TOP_P_KEY = "top-p";
+  public static final double TOP_P_DEFAULT = 0.9;
+  public static final String TOKENIZER_KEY = "tokenizer";
```
Comment (bilbo7833) [inline java/apirag-core/src/main/java/com/datasqrl/ai/models/AbstractModelConfiguration.java:19]: @mbroecheler Why should we make the tokenizer customizable? Most LLMs require a specific tokenizer and we choose the right one in our model type. I would not add this to the first version.

File: java/apirag-groq/src/main/java/com/datasqrl/ai/models/groq/GroqTokenCounter.java [hunk 5]
```diff
@@ -34,7 +34,7 @@ public int countTokens(FunctionDefinition function) {
     return countTokens(jsonString);
   }
 
-  public static GroqTokenCounter of(GroqChatModel model) {
-    return new GroqTokenCounter(HuggingFaceTokenizer.newInstance(model.tokenizerName));
+  public static GroqTokenCounter of(GroqModelConfiguration model) {
```
Comment (bilbo7833) [inline java/apirag-groq/src/main/java/com/datasqrl/ai/models/groq/GroqTokenCounter.java:37]: Why go through the configuration, since we have a `GroqModelType` here? Is this on purpose, so that we always interact with the Configuration instead of the ModelType?

File: java/apirag-core/src/main/java/com/datasqrl/ai/models/AbstractModelConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+@AllArgsConstructor
+public abstract class AbstractModelConfiguration implements ModelConfiguration {
+
+  public static final String MODEL_NAME_KEY = "name";
+  public static final String MAX_INPUT_TOKENS_KEY = "max_input_tokens";
+  public static final String MAX_OUTPUT_TOKENS_KEY = "max_output_tokens";
+  public static final String TEMPERATURE_KEY = "temperature";
+  public static final double TEMPERATURE_DEFAULT = 0.7;
+  public static final String TOP_P_KEY = "top-p";
```
Comment (bilbo7833) [inline java/apirag-core/src/main/java/com/datasqrl/ai/models/AbstractModelConfiguration.java:17]: @mbroecheler Shall we try to stick to snake_case for the configuration? Most LLMs are configured so. So we might want to unify our configs to use underscores throughout. Wdyt? `top_p`

Comment (mbroecheler) [reply java/apirag-core/src/main/java/com/datasqrl/ai/models/AbstractModelConfiguration.java:17]: Yes, good eye!

**Review by bilbo7833 — APPROVED** [review]
I have implemented the configuration changes for Bedrock. Now, I will rebase the Google Vertex branch on top of this branch, so that I can also implement the same configuration options there

File: html/bot.svg
Comment (bilbo7833) [inline html/bot.svg:1]: Why did you delete the chatbot?
